### PR TITLE
[NP-3680] - ucjDAO fileDAO Sync

### PR DESCRIPTION
**Pull request by nanopay-arthur**
Ref: https://nanopay.atlassian.net/browse/NP-3680

### CODE REVIEW

File: src/foam/nanos/crunch/ClientCrunchService.js [hunk 1]
```diff
@@ -9,15 +9,138 @@ foam.CLASS({
...
+          this.logger.error("getJunction", capabilityId, e);
+
+          // On failure, report that the capability is available
+          var ucj = await this.buildAssociatedUCJ(x, capabilityId, this.subject);
+          return ucj;
+        }
+      }
+    },
+    {
+      name: 'updateJunction',
```
Comment (kgrgreer) [inline src/foam/nanos/crunch/ClientCrunchService.js:37]: Isn't this the only method we need? What are the others for?

**Review by KolombetM — APPROVED** [review]

File: src/foam/nanos/crunch/ClientCrunchService.js (hunk 1 above)
Comment (nanopay-arthur) [reply src/foam/nanos/crunch/ClientCrunchService.js:37]: I assumed we needed to define the other CrunchService methods updateJunction calls within itself, but now that you mention it, I don't see why client's updateJunction method can't just call ServerCrunchService's getJunction method. Will update.

File: src/foam/nanos/fs/FileArrayDAODecorator.js [hunk 2]
```diff
@@ -25,42 +25,51 @@ foam.CLASS({
...
-      });
+    async function processFiles(obj) {
+      var props1 = obj.cls_.getAxiomsByClass(foam.nanos.fs.FileArray);
+      var values = await Promise.all(props1.map(prop => Promise.all(prop.f(obj).map(f => this.processFile(f)))));
+      values.forEach(f => f.forEach(f2 => {
+        f2.dataString = undefined;
+        f2.data = undefined;
+      }));
+      props1.forEach((prop, i) => prop.set(obj, values[i]));
+      var props2 = obj.cls_.getAxiomsByClass(foam.core.FObjectProperty).filter((p) => ! foam.dao.DAOProperty.isInstance(p));
```
Comment (jlhughes) [inline src/foam/nanos/fs/FileArrayDAODecorator.js:54]: Question: is a FileArray also a FObjectProperty? Curious if obj FileArray properties are processed twice?

Comment (nanopay-arthur) [reply src/foam/nanos/fs/FileArrayDAODecorator.js:54]: Yes, currently a nested FileArray would be processed twice... We can add another condition to the filter of props2 to be ! props1.includes(p).

Comment (nanopay-arthur) [reply src/foam/nanos/fs/FileArrayDAODecorator.js:54]: actually, I don't think my suggestion makes sense...

Comment (nanopay-arthur) [reply src/foam/nanos/fs/FileArrayDAODecorator.js:54]: actually, it does.

**Pull request merged**